sys/src/9/amd64/devvq.c

```diff
@@ -0,0 +1,595 @@
+			return -1;
+		char *dmap = mapdev(cvq[s]);
+		snprint(up->genbuf, sizeof up->genbuf, "%s-%d", dmap?dmap:"virtio", s);
+		q = (Qid) {QID(s, Qdevtop), 0, QTDIR};
+		devdir(c, q, up->genbuf, 0, eve, DMDIR|0555, dp);
+		return 1;
+	case Qdevtop:
+		vdidx = DEV(c->qid);
+		if(vdidx >= nvq)
+			return -1;
+		if(s == 0) {					// device configuration area, report as a read-only file
+			snprint(up->genbuf, sizeof up->genbuf, "devcfg");
+			q = (Qid) {QID(vdidx, Qdevcfg), 0, 0};
+			devdir(c, q, up->genbuf, cvq[vdidx]->dcfglen, eve, 0444, dp);
+			if(cvq[vdidx]->dcmtime == -1)			// initialize lazily because at the moment of the driver bring-up
+				cvq[vdidx]->dcmtime = seconds();	// seconds() does not seem to return correct value
+			dp->mtime = cvq[vdidx]->dcmtime;
+			return 1;
+		} else {						// virtqueues: odd s yield data entries, even s yield control entries
+			int t = s - 1;
+			if(t >= cvq[vdidx]->nqs * 2)
+				return -1;
+			int nq = (t & ~1) >> 1;
+			q = (Qid) {VQQID(nq, vdidx, (t&1)?Qvqctl:Qvqdata), 0, 0};
+			snprint(up->genbuf, sizeof up->genbuf, "vq%03d.%s", nq, (t&1)?"ctl":"data");
+			devdir(c, q, up->genbuf, (t&1)?0:cvq[vdidx]->vqs[nq]->num, eve, (t&1)?0222:0666, dp);
+			return 1;
+		}
+	default:
+		return -1;
+	}
+	return -1;
+}
+
+
+static Chan*
+vqattach(char *spec)
+{
+//print("vqattach %s\n", spec);
+	return devattach(vqdevtab.dc, spec);
+}
+
+
+Walkqid*
+vqwalk(Chan* c, Chan *nc, char** name, int nname)
+{
+//print("vqwalk %d -> %d\n", c?TYPE(c->qid):-1, nc?TYPE(nc->qid):-1);
+	return devwalk(c, nc, name, nname, (Dirtab *)0, 0, vqgen);
+}
+
+static int32_t
+vqstat(Chan* c, uint8_t* dp, int32_t n)
+{
+//print("vqstat %d\n", TYPE(c->qid));
+	return devstat(c, dp, n, (Dirtab *)0, 0L, vqgen);
+}
+
+// A process holding open the device configuration area has access to its virtqueues.
+// Other processes will be denied any access. PID of the claiming process is stored
+// in the ownpid field of the device descriptor. An unclaimed device has -1 in it.
+
+static Chan*
+vqopen(Chan *c, int omode)
+{
+//print("v9open %d\n", TYPE(c->qid));
+	Proc *up = externup();
+	uint t = TYPE(c->qid);
+	uint vdidx = DEV(c->qid);
+	if(vdidx >= nvq)
+		error(Ebadarg);
+	vqctl *vc = cvq[vdidx];
+	switch(t) {
+		case Qdevcfg:
+			if(vc->ownpid != -1)
+				error(Eperm);
+			vc->ownpid = up->pid;
+		case Qvqdata:
+		case Qvqctl:
+			if(vc->ownpid != up->pid)
+				error(Eperm);
+			break;
+		default:
+			break;
+	}
+	c = devopen(c, omode, (Dirtab*)0, 0, vqgen);
+	switch(t) {
+	default:
+		break;
+	}
+	return c;
+}
+
+static void
+vqclose(Chan* c)
+{
+//print("vqclose %d\n", TYPE(c->qid));
+	Proc *up = externup();
+	uint t = TYPE(c->qid);
+	uint vdidx = DEV(c->qid);
+	if(vdidx >= nvq)
+		return;
+	vqctl *vc = cvq[vdidx];
+	if (vc->ownpid != up->pid)
+		return;
+	switch(t) {
+		case Qdevcfg:
+			vc->ownpid = -1;
+			break;
+		default:
+			break;
+	}
+}
+
+
+
+// Reading from the device configuration area is byte-by-byte (8bit ports) to avoid dealing with
```

Comment:
we can talk about this later but this is going to be a huge perfomance hit. Each vmexit will require emulation on the part of the host, and that won't be cheap. I think this will be a performance issue.